grakn-test-tools/src/main/java/ai/grakn/engine/rpc/SynchronousObserver.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Grakn - A Distributed Semantic Database
+ * Copyright (C) 2016-2018 Grakn Labs Limited
+ *
+ * Grakn is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Grakn is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Grakn. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ */
+
+package ai.grakn.engine.rpc;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+import io.grpc.stub.StreamObserver;
+
+import javax.annotation.Nullable;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+/**
+ * Wrapper for synchronous bidirectional streaming communication - i.e. when there is a stream of {@link Request}s and
+ * a stream of {@link Response}s.
+ *
+ * A request is sent with the {@link #send(Request)}} method, and you can block for a response with the
+ * {@link #receive()} method.
+ *
+ * <pre>
+ * {@code
+ *
+ *     try (SynchronousObserver<TxRequest, TxResponse> tx = SynchronousObserver.create(stub::tx) {
+ *         tx.send(openMessage);
+ *         TxResponse doneMessage = tx.receive().elem();
+ *         tx.send(commitMessage);
+ *         Throwable validationError = tx.receive.throwable();
+ *     }
+ * }
+ * </pre>
+ *
+ * @author Felix Chapman
+ *
+ * @param <Request> The type of requests being received
+ * @param <Response> The type of responses being sent
+ */
+public class SynchronousObserver<Request, Response> implements AutoCloseable {
```

Comment:
Lets chat about this one in person. I think I understand it but I want to double check. I think this is used to convert the asynchronous conversation into a synchronous one. I just want to check how.